octodns/provider/googlecloud.py

```diff
@@ -0,0 +1,343 @@
+            derived from it. calls .create() method and returns it.
+
+            :param dns_name: fqdn of zone to create
+            :type  dns_name: str
+
+            :type return: new google.cloud.dns.ManagedZone
+        """
+        # Zone name must begin with a letter, end with a letter or digit,
+        # and only contain lowercase letters, digits or dashes
+        zone_name = re.sub("[^a-z0-9-]", "",
+                           dns_name[:-1].replace('.', "-"))
+        # make sure that the end result did not end up wo leading letter
+        if re.match('[^a-z]', zone_name[0]):
+            # I cannot think of a situation where a zone name derived from
+            # a domain name would'nt start with leading letter and thereby
+            # violate the constraint, however if such a situation is
+            # encountered, add a leading "a" here.
+            zone_name = "a%s" % zone_name
+
+        gcloud_zone = self.gcloud_client.zone(
+            name=zone_name,
+            dns_name=dns_name
+        )
+        gcloud_zone.create(client=self.gcloud_client)
+
+        self.log.info("Created zone %s. Fqdn %s." %
+                      (zone_name, dns_name))
+
+        return gcloud_zone
+
+    def _get_gcloud_records(self, gcloud_zone, page_token=None):
+        """ Generator function which yields ResourceRecordSet for the managed
+            gcloud zone, until there are no more records to pull.
+
+            :param gcloud_zone: zone to pull records from
+            :type gcloud_zone: google.cloud.dns.ManagedZone
+            :param page_token: page token for the page to get
+
+            :return: a resource record set
+            :type return: google.cloud.dns.ResourceRecordSet
+        """
+        gcloud_iterator = gcloud_zone.list_resource_record_sets(
+            page_token=page_token)
+        for gcloud_record in gcloud_iterator:
+            yield gcloud_record
+        # This is to get results which may be on a "paged" page.
+        # (if more than max_results) entries.
+        if gcloud_iterator.next_page_token:
+            for gcloud_record in self._get_gcloud_records(
+                    gcloud_zone, gcloud_iterator.next_page_token):
+                # yield from is in python 3 only.
+                yield gcloud_record
+
+    def _get_gcloud_zone(self, dns_name, page_token=None, create=False):
+        """Return the ManagedZone which has has the matching dns_name, or
+            None if no such zone exist, unless create=True, then create a new
+            one and return it.
+
+            :param dns_name: fqdn of dns name for zone to get.
+            :type dns_name: str
+            :param page_token: page token for the page to get
+            :type page_token: str
+            :param create: if true, create ManagedZone if it does not exist
+                           already
+
+            :type return: new google.cloud.dns.ManagedZone
+        """
+        # Find the google name for the incoming zone
```

Comment:
I think this may have the same 2-page problem as records. It'd also be O(n^2) since it'd walk all zones for each managed zone to find the one we're after. It'd probably be best to break up the process of getting the full list of zones from finding the one we're after. 

The method that gets all the zones could just store them in a _cached hash of zone -> object. There's several providers that do that already. Otherwise the code would look pretty much the same as the `while True` loop for records. 

`_get_gcloud_zone` would then just be able to look up the zone in the cached hash.

Reply:
Yes I was thinking this very same thing. This would also ensure that when a new zone is created, it's possible to check that the generated zone name is not already in use by some other zone.